### DISCUSSION TITLE
Forth in Java

### DISCUSSION THREAD

**Discussion by ruv** [ForthHub/discussion#132] [General]
Could you suggest a Forth implementation that can be embedded in a Java application?
So actually it is not obligated that it's in Java, it can be in any language that compiles in JVM.

Also, it shall have some FFI to Java methods, and be able to catch exceptions in Java methods.

**Comment by SirWumpus** [comment]
What about a C implementation with JNI wrappers?

**Reply by ruv** [reply]
Interesting. But then it cannot be embedded in a Java application, but vise versa — it calls the Java application.

Anyway, could you suggest one having JNI with bells and whistles out of the box?

So, something like:
```
 "test" org.foo.bar . \ prints -1
```
just works?

**Reply by SirWumpus** [reply]
Well its been sometime since I worked with JNI, but I once wrote a cyber cafe time management system in Java with JNi wrappers to Windows API functions to control access to the registry, full screen windows, and stuff etc.  Wrapping a handful of C function to hook into Java should be easy.

I've written [post4](https://github.com/SirWumpus/post4) with the design already in mind to make it a linkable interpreter with other projects, not just a stand alone CLI.  It shouldn't be too hard to make a library callable from Java and propagate exceptions up to Java.

**Reply by SirWumpus** [reply]
The only issue with C based version is that interpreter library would be host dependant, unlike something built on top of the JVM.

**Reply by ruv** [reply]
@SirWumpus, thank you for sharing your project. I would suggest adding the [`Forth-2012` tag](https://github.com/topics/forth-2012).

**Reply by SirWumpus** [reply]
Thanks. Tag added. I had been putting off sharing it until it was well along with the test suite.

**Reply by SirWumpus** [reply]
Today I got basic JNI support implemented in [post4](https://github.com/SirWumpus/post4)  (not yet pushed to GitHub) where Java can evaluate a string or file, enter the REPL, and receive exceptions.  Its pretty one way just now and needs some polish.

However thinking on the reverse, being able to call Java methods from Forth, raises a few issues to think about:

* Object construction and passing `this` back to Java.
* Object life cycle. Forth needs to ensure objects it creates, via Java constructors, and only it knows about are not GC'ed behind its back. How to release an object (reference) when done.
* Overloaded constructors and methods.  In most cases the method with the greatest "arity` (think Erlang) is probably the ideal one to invoke, but it assumes those with lesser arity are convenience methods.
* Java package and method names are some times stupidly long, which might affect input buffer size.
* I/O redirection, let Java feed input, capture output.
* For file and string evaluation, are results from the data and/or float stacks expected to be passed back to Java? How deep?

**Reply by ruv** [reply]
> Object life cycle.

As I understand, an object is release (freed) when it is not referenced anymore. So, to release an object, it's need to remove the references to this object.

I see several approaches to access Java objects in Forth:

1. Make the Forth data stack a stack of Java objects (if a Forth system is embedded in Java). So, all items on the data stack are boxed (in terms of Java). Some operations can be not applicable to objects (e.g., store into memory by means of `!`). Then it means, an object reference is not a [subtype](https://forth-standard.org/standard/usage#usage:type) of `x` or `i*x` in Forth.

2. Introduce an additional object stack and object variables for Java objects. When an object is dropped from the object stack, and it is not saved into a variable, then we don't have a direct reference to this object in Forth. If it isn't referenced by other objects too, JVM can destroy it.

3. Introduce an internal array of objects. The indices in this array represent Java objects in Forth (avoid number 0 in Forth to represent any object). In this approach an object should be released explicitly in Forth, and the corresponding reference is removed from the array by this operation.

> Overloaded constructors and methods. In most cases the method with the greatest "arity` (think Erlang) is probably the ideal one to invoke, but it assumes those with lesser arity are convenience methods.

When a method name cannot be resolved without specifying its arity, the arity can be specified as an argument of a _dlsym_-like operation.  A more difficult case is when the type of arguments should be also specified to resolve a method (I don't know whether it's a case in Java reflection API).

Probably, a lazy way is to recognize something like `java:+(int,int)`. The information about arity and data types is also needed to proper marshal parameters from Forth to a Java method. If this information is not retrieved via introspection, then it should be specified in Forth (like in this example).

**Reply by SirWumpus** [reply]
> >  Object life cycle.

> As I understand, an object is release (freed) when it is not referenced anymore. So, to release an object, it's need to remove the references to this object.

Yep.  I just need to read up more to make sure I understand possible JNI gotchas.  As to how to manage it, I'll review your suggestions (after I get my brain out of 1st gear).  I'm sure something will jump out from a dark corner and go "boo!" (just like learning to fly HHGTTG).

> When a method name cannot be resolved without specifying its arity, the arity can be specified as an argument of a dlsym-like operation. A more difficult case is when the type of arguments should be also specified to resolve a method (I don't know whether it's a case in Java reflection API).

Ya.  This is the case with Java, method name "foo" and signature, eg. "(Ljava.lang.String;[J[D)V" (fictitious example), is essentially 
```
void foo(String, long[], double[])
```
Constructors are simply methods called "`<init>`".  See [the JNI](https://docs.oracle.com/en/java/javase/17/docs/specs/jni/index.html) `GetFieldID` and `GetMethodID` if curious.

I've been looking at how to get (other than using `javap -s obj.path.name` CLI) the list of methods and signatures.  Documentation is scattered.  Found one [tit-bit](https://docs.oracle.com/javase/tutorial/reflect/member/methodType.html) late yesterday.

>> For file and string evaluation, are results from the data and/or float stacks expected to be passed back to Java? How deep?

As for this question, I implemented a basic solution that allows Java to examine the data and float stacks on demand. Pushed work to date on [Post4](https://github.com/SirWumpus/post4) (this is the 1st half Java to Forth aspect (the easy part)).

**Reply by ruv** [reply]
In **8th** (a Forth-like language), the use of FFI to Java (on Android) [looks](https://8th-dev.com/manual.html#java-interface-(android-only)) as follows:

```
\ First get the class on the stack:
"java/lang/Thread" jclass
\ And now make a method item:
"sleep" "(J)V" jmethod
\ And finally, invoke the method:
[ 200 ] swap jcall
```

The words `jclass`, `jmethod`, `jcall` are described in [Words list](https://8th-dev.com/words.html).

**Reply by SirWumpus** [reply]
Interesting.  But essentially "8th" is doing all the low-level grunt work the JNI library should do.  Your original ask for something like:
```
S" hello terra" System.out.println 
123.45 System.out.println
```
would be nice; have the some means (JNI, Forth, other) that does all the prep. work for the call.  Maybe a hybrid, where one supplies the signature at least might help select which variant, but that has issues too I think as signatures get long.
```
S" hello terra" S" (Ljava.lang.String)V" System.out.println 
123.45 S" (D)V" System.out.println
```
- - -

It might help to understand your objective better.  Is it to take advantage of Java threading to run multiple Forth instances, so only some subset is needed.  Or more general purpose as in "this would be cool to do".  Another concern I have is that beyond Java's primitive types, there's a lot of work to setup / teardown objects, arrays, and the like for arguments or results.  Sure words can be created to manage some of that, but Forth (in my mind) works best in small spaces and micro manages a task for space and efficiency, but Java is a huge high level monster with many packages for assorted needs.  

What does the fusion of Forth and Java bring that can't be done just in Java alone?  What is the bigger picture?

**Reply by ruv** [reply]
>  so only some subset is needed.

No, Java threading is just an accidental example from **8th**.

> What does the fusion of Forth and Java bring that can't be done just in Java alone?

I think, it's a question of convenience, not possibility.

The fusion of Forth and Java can be implemented in two variants:

1. Forth is integrated into a Java application (Java calls Forth). It can bring to the application (I suppose):
    - an interactive terminal (REPL) with introspection/debugging facilities,
    - testing, configuring, scripting (without recompilation of the application).

4. Access to Java from a Forth system (Forth calls Java). This brings capabilities to:
    - test/introspect/debug Java components (classes),
        - partially, what [JShell](https://docs.oracle.com/javase/10/jshell/introduction-jshell.htm) now provides.
    - use libraries/components from the huge Java ecosystem in a Forth program,
    - create light wrappers/scripts over a Java component just in Forth (without creating a Java application for that),
        - as an example, once I needed a simple wrapper over [Apache™ FOP](https://xmlgraphics.apache.org/fop/) component.
        - (I consider Forth far simpler than Java)

**Reply by ruv** [reply]
> Your original ask for something like:
> ```
> S" hello terra" System.out.println 
> 123.45 System.out.println
> ```
I guess, it should not work in the second case. Forth does not know whether it's a string length or just a number. So, only a string should be passed to this method.

By the way, if there is only one method of a given name, it can be resolved without providing signature, at least via enumeration of all the methods. Example: [ReflectionMachine.java](https://github.com/Reschivon/Forth-2020-Asteroids-Demo/blob/master/src/Forth/ReflectionMachine.java) (in [Forth-2020-Asteroids-Demo](https://github.com/Reschivon/Forth-2020-Asteroids-Demo))

> one supplies the signature at least might help select which variant, but that has issues too I think as signatures get long.
> ```
> S" hello terra" S" (Ljava.lang.String)V" System.out.println 
> ```

If it's needed to specify a signature in some case, it's a good thing to be able to specify it as an immediate argument, e.g.:
```
System.out.println((Ljava.lang.String)V)
```
— for better performance of compiled code (and simpler implementation of that).

Concerning long signatures — if they used several times, it can be wrapped into a simple definitions as:
```
: println ( sd.text -- ) System.out.println((Ljava.lang.String)V) ;
```

**Reply by SirWumpus** [reply]
> By the way, if there is only one method of a given name, it can be resolved without providing signature.

From what I've read and used so far of the JNI this past weekend, get/set fields, and method calls, even if there is only one by that name still require a signature.  Its a two step process: first lookup a jfieldID or jmethodID based on the class, then use that ID to interact with the field or method of an object. 

I've not looked at 8th at all, but I suppose to handle overloading, one would prepare some helper words;
```
: sysout-println ( caddr u -- )
  S" (Ljava.lang.String)V" S" System.out.println" jcall
;
: sysout-double-println ( float -- )
  S" (D)V" S" System.out.println" jcall
;
: some-java-obj ( xn ... xi n this -- x )
  >R set-jarray >R S" ([I)J" S" some.java.Woot.method" jcall
;
```
The above is just ideas off the top of my head.  

Just noticed your edit about wrappers; had similar idea in mind.  Maybe not as nice as originally envisioned, but certainly workable -- put all the wrappers in files and include them, similar to Java `import java.lang.System.*`.   I'll futz with it some more.

**Reply by ruv** [reply]
> From what I've read and used so far of the JNI this past weekend, get/set fields, and method calls, even if there is only one by that name still require a signature.

Yes. But then what can prevent us to use the Reflection API to enumerate all the methods of a class?

Namely, we can use [`CallObjectMethod`](https://docs.oracle.com/en/java/javase/19/docs/specs/jni/functions.html#calltypemethod-routines-calltypemethoda-routines-calltypemethodv-routines) to call [`java.lang.Object.getClass`](https://docs.oracle.com/en/java/javase/19/docs/api/java.base/java/lang/Object.html#getClass()), [`java.lang.Class.getMethods`](https://docs.oracle.com/en/java/javase/19/docs/api/java.base/java/lang/Class.html#getMethods()), [`java.lang.reflect.Method.getGenericParameterTypes`](https://docs.oracle.com/en/java/javase/19/docs/api/java.base/java/lang/reflect/Method.html#getGenericParameterTypes()), etc, and use [`FromReflectedMethod`](https://docs.oracle.com/en/java/javase/19/docs/specs/jni/functions.html#reflection-support) to get the _jmethodID_ for a method. 

See also an [example](https://stackoverflow.com/questions/58130302/finding-if-a-java-class-is-final-in-jni-by-using-reflection) on Stackoverflow.

**Reply by SirWumpus** [reply]
@ruv So I've finally implemented an initial set of words for [Forth to Java support](https://github.com/SirWumpus/post4/tree/master/jni) in Post4, which are pretty much JNI wrappers.  I still haven't seen how reflection could help just yet.  

The documentation is light for the moment (as it might change), but there are some examples of how these might be used, eg. [java.lang.System.p4](https://github.com/SirWumpus/post4/blob/master/jni/java.lang.System.p4) and [java.lang.Thread.p4](https://github.com/SirWumpus/post4/blob/master/jni/java.lang.Thread.p4).

I did look at the three "8th" words for Android Java support (which seems limited) and suspect they hide (punt) a lot of the object reference management, no field get/set, box/unbox strings or arrays, and more.

**Comment by MitchBradley** [comment]
My [CForth](https://github.com/MitchBradley/cforth) could once run on a Java-based phone.  It worked by transliterating C to Java with some preprocessor trickery.  To make that work, I had to change all the pointer arithmetic to array references.  The java code is still there, but has not been maintained in years so I would be really surprised if it still works without some tweaking.

**Reply by ruv** [reply]
The most difficult part (which is absent as I can see), is an FFI. It should use the Java reflection to resolve an actual method (from a name) and its parameters, and properly pass (or marshal) them from the data stack to the method, and pass the result back to the data stack.

**Reply by MitchBradley** [reply]
The base CForth has a reasonable facility for calling C functions. You have to specify the parameters, and it depends on the fact that C essentially passes everything on the call stack as a glorified int.  Doing an automated FFI for Java would be possible, I suppose, but it probably would be a lot of work, as you surmise.

My old company FirmWorks did a Java implementation of Open Firmware for Sun at one point, but I can't find the code.

**Reply by ruv** [reply]
I just think that this work (FFI for Java) was already done a number of times, and there is no need to do it again -)

**Comment by jwoehr** [comment]
You might be able to fiddle with my ancient FIJI ForthIsh Java Interpreter
<https://sourceforge.net/projects/fiji/> which is an object oriented Forth
in Java that can call into Java.

**Reply by ruv** [reply]
@jwoehr, many thanks! I even see an updated fork: https://github.com/bluekelp/fiji-core (2018)
At the first glance it looks like what I asked.

**Comment by andrewtholt** [comment]
Others have probably mention previous forth in java implementations, it
can't hurt

Somewhat old now but it may be of interest:

https://fiji.sourceforge.net/

I vaguely remember calling methods in java classes from forth like words.
It might be worth a look

HTH,
Andrew

**Comment by jwoehr** [comment]
3. Write Forth-like language in Java that can call Java.
https://sourceforge.net/projects/fiji/

On Wed, Feb 22, 2023 at 9:37 AM ruv ***@***.***> wrote:

> The fusion of Forth and Java can be implemented in two variants:
>
>    1.
>
>    Forth is integrated into a Java application (Java calls Forth). It can
>    bring to the application (I suppose):
>    - an interactive terminal (REPL) with introspection/debugging
>       facilities,
>       - testing, configuring, scripting (without recompilation of the
>       application).
>    2.
>
>    Access to Java from a Forth system (Forth calls Java). This brings
>    capabilities to:
>    - test/introspect/debug Java components (classes),
>          - partially, what JShell
>          <https://docs.oracle.com/javase/10/jshell/introduction-jshell.htm>
>          now provides.
>       - use libraries/components from the huge Java ecosystem in a Forth
>       program,
>       - create light wrappers/scripts over a Java component just in Forth
>       (without creating a Java application for that),
>          - as an example, once I needed a simple wrapper over Apache™ FOP
>          <https://xmlgraphics.apache.org/fop/> component.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ForthHub/discussion/discussions/132#discussioncomment-5079223>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABDEBBDQIDIIA5FIRI3S5ZDWYY6FXANCNFSM6AAAAAAUFQY3NA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 

*JACK WOEHR, IBM Champion 2021-2023
<https://www.credly.com/users/jack-woehr/badges>*

*SENIOR IBM i ENGINEER*

  303.847.8442
  ***@***.***
  stay connected
<https://www.linkedin.com/company/absolute-performance-inc./>

<https://www.absolute-performance.com/>
 www.absolute-performance.com

NON-DISCLOSURE NOTICE:  This communication including any and all
attachments is for the intended recipient(s) only and may contain

confidential and privileged information.  If you are not the intended
recipient of this communication, any disclosure, copying further

distribution or use of this communication is prohibited.  If you received
this communication in error, please contact the sender and

delete/destroy all copies of this communication immediately.
